[AN-Issue-1347] Implemented Automated transaction execution flow

aregng opened the pull request.

Commit:
[AN-Issue-1347] Implemented Automated transasction execution flow

sjoshisupra commented on aptos-move/framework/supra-framework/sources/transaction_validation.move:
```diff
@@ -167,6 +167,35 @@ module supra_framework::transaction_validation {
+        _script_hash: vector<u8>,
+    ) {
+        let gas_payer = signer::address_of(&sender);
+        assert!(
+            timestamp::now_seconds() < txn_expiration_time,
+            error::invalid_argument(PROLOGUE_ETRANSACTION_EXPIRED),
+        );
```
I was under the impression that if an automation task is beyond it's expiry, the SMR layer would never include it in the set of tx to be executed. Is that still correct? Because otherwise it would result in continuous failed assert here in the prologue.

aregng replied:
That is true, but as mentioned in my previous comment, I think each layer should validate its input and not rely on the other component correctness. If we do not agree with this approach and here we will always assume that input will not be expired, then this check can be removed.

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/transaction_validation.move:
```diff
@@ -167,6 +167,35 @@ module supra_framework::transaction_validation {
+        txn_gas_price: u64,
+        txn_max_gas_units: u64,
+        txn_expiration_time: u64,
+        _script_hash: vector<u8>,
```
In the old prologue functions, I noticed that we pass chain_id and validate it using an assertion, such as `assert!(chain_id::get() == chain_id, error::invalid_argument(PROLOGUE_EBAD_CHAIN_ID))`.
Do we need this validation in the `automated_transaction_prologue`?

BTW, there is a typo in `automated_transaction_prologue` — `tranaction`.

aregng replied:
Thanks for the hint. also _script_hash should be removed from here.
as for chain-id, it can be there or not.
`AutomatedTransaction` will be constructed internally by `smr-moonshot-executor` component, and the chain ID will be set there. so theoretically it should be always valid.
But as long as `smr-moonshot-executor` and `MoveVM`  are completly isolated components, and one provides input to the other, in my understanding, each component should validate its input and not rely on the fact that the other one will not produce any buggy information. This way the if there is a bug in one layer which is not tested, then the other layer can at least point it out.
But we need to agree whether we want to go with this approach or not.

sjoshisupra reviewed (commented):
@aregng , it seems that from the issue [1347](https://github.com/Entropy-Foundation/smr-moonshot/issues/1347) I did not see 
- Chain ID check being done (if it is not done for other tx, we may not do it for automation as well)
- Check that there is a corresponding _active_ task for this tx (each automation task would have a unique id provided by registry) and sender seems missing. If we assume however that nodes are _honest_ I wonder if this check is required though.

aregng commented:
> @aregng , it seems that from the issue [1347](https://github.com/Entropy-Foundation/smr-moonshot/issues/1347) I did not see
> 
> * Chain ID check being done (if it is not done for other tx, we may not do it for automation as well)
> * Check that there is a corresponding _active_ task for this tx (each automation task would have a unique id provided by registry) and sender seems missing. If we assume however that nodes are _honest_ I wonder if this check is required though.

- ChainID, During pervious discussion it was marked as optional, that's why I have not add it and wanted to clarify one more time. The check  is done for user-transaction. So I will add it here as well.
- Regarding check against registry: As mentioned in the above comment, either we always trust that the input provided by other component can not be incorrect, or each component does some checks  on its own as well. 
Can you point where sender is missing ?
@aregng I am referring to point 3 in prologue requirement of 1347 (https://github.com/Entropy-Foundation/smr-moonshot/issues/1347), I believe it is not checked in the prologue that task-id indeed belong to the sender.

Commit:
Updated automated task prologue and added e2e tests

- Added chain-id verification and task availability check by task index
- Added e2e tests for AutoamtedTransaction

Commit:
Fixed test failure

sjoshisupra commented on aptos-move/aptos-gas-schedule/src/gas_schedule/aptos_framework.rs:
```diff
@@ -269,6 +270,7 @@ crate::gas_schedule::macros::define_gas_parameters!(
 
         [transaction_context_get_txn_hash_base: InternalGas, { 10.. => "transaction_context.get_txn_hash.base" }, 735],
         [transaction_context_get_script_hash_base: InternalGas, "transaction_context.get_script_hash.base", 735],
+        [transaction_context_get_txn_app_hash_base: InternalGas, "transaction_context.get_txn_app_hash.base", 735],
```
Do we need this anymore? Since native function would no longer be needed I think.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/genesis.move:
```diff
@@ -188,6 +189,7 @@ module supra_framework::genesis {
         reconfiguration::initialize(&supra_framework_account);
         block::initialize(&supra_framework_account, epoch_interval_microsecs);
         state_storage::initialize(&supra_framework_account);
+        automation_registry::initialize(&supra_framework_account);
```
I don't think we should call this here. Otherwise replay of the chain would not agree on its state. Since we have launched mainnet, this should not be touched now. `automation_registry::initialize` can be called via governance script (upgrade framework followed by call to initialize)

aregng replied:
The testing will not be easy if we do not have initialize called here. as during testing we start the chain from genesis.
I remember you once mentioned that you want to avoid conditional initialization, but at this moment I do not see easy way to have the module initialized properly with genesis startup.
Featuer enable/disable interface might be investigated to picture and cover all scenarios.

aregng replied:
Mentioned in ticket https://github.com/Entropy-Foundation/smr-moonshot/issues/1422 to be addressed later.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/stake.spec.move:
```diff
@@ -437,7 +437,8 @@ spec supra_framework::stake {
     spec on_new_epoch {
         pragma verify = false; // TODO: set because of timeout (property proved).
-        pragma disable_invariants_in_body;
+        // TODO: Why we need to disable invariants in body? (SUPRA)
```
I think Move prover will fail compilation if we do not have this pragma.

aregng replied:
This is not part of my change-set. it comes from this change-set f9652d1f0472fc60f887605ff13f779801b039e4

sjoshisupra commented on aptos-move/framework/supra-stdlib/sources/enumerable_map.move:
```diff
@@ -0,0 +1,276 @@
+    ): KeyValue<K, V> {
+        assert!(contains(map, key), error::not_found(EKEY_ABSENT));
+        table::borrow_mut(&mut map.map, key).value = new_value;
+        KeyValue { key, value: new_value }
```
@nizam-supraoracles why should it return `KeyValue` with `new_value`? A user who called it would already know the answer. It may be useful if the returned value is `KeyValue { key, old_value}` change the comments also accordingly.

aregng replied:
will be addressed in scope https://github.com/Entropy-Foundation/smr-moonshot/issues/1425 ticket

sjoshisupra commented on aptos-move/framework/supra-stdlib/sources/enumerable_map.move:
```diff
@@ -0,0 +1,276 @@
+    }
+
+    /// Add Multiple Keys in the Enumerable Map
+    public fun add_value_bulk<K: copy+drop, V : store+drop+copy>(
```
Just like `vector::for_each` and `vector::for_each_ref`, it may be helpful if we add `enumerable_map::for_each_value` and `enumerable_map::for_each_value_ref`

aregng replied:
will be addressed in scope https://github.com/Entropy-Foundation/smr-moonshot/issues/1425 ticket

sjoshisupra commented on ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs:
```diff
@@ -835,6 +859,25 @@ pub fn convert_transaction(
         },
+        Transaction::AutomatedTransaction(at) => {
+            timestamp = Some(convert_timestamp_usecs(at.timestamp.0));
+            let expiration_timestamp_secs = Some(convert_timestamp_secs(std::cmp::min(
+                at.meta.expiration_timestamp_secs.0,
+                chrono::NaiveDateTime::MAX.timestamp() as u64,
+            )));
```
@aregng I am not sure I quite understand what is happening here. `min` of `MAX` time and `at.meta.expiration_timestamp_secs.0` would always result in `at.meta.expiration_timestamp_secs.0`, right? Not sure what is happening here.

aregng replied:
My assumption will be that if user specified timestamp exceeds expected MAX of the  `chrono::NaiveDateTime::MAX.timestamp()` it will take the later one. This part might require revision as AutoamtedTransactions will never be posted by users. I need to spend more time to understand APTOS-API and how it is utilized to make sure that these changes were valid or invalid. But at this moment they do not play any role  it the main automation-flow implementation.

`chrono` uses `i64` for timestamp and  we use `u64` so `i64`::MAX can be potentially less than `u64` value specified for expiration time.

aregng commented:
> > @aregng , it seems that from the issue [1347](https://github.com/Entropy-Foundation/smr-moonshot/issues/1347) I did not see
> > 
> > * Chain ID check being done (if it is not done for other tx, we may not do it for automation as well)
> > * Check that there is a corresponding _active_ task for this tx (each automation task would have a unique id provided by registry) and sender seems missing. If we assume however that nodes are _honest_ I wonder if this check is required though.
> 
> * ChainID, During pervious discussion it was marked as optional, that's why I have not add it and wanted to clarify one more time. The check  is done for user-transaction. So I will add it here as well.
> * Regarding check against registry: As mentioned in the above comment, either we always trust that the input provided by other component can not be incorrect, or each component does some checks  on its own as well.
>   Can you point where sender is missing ?
>   @aregng I am referring to point 3 in prologue requirement of 1347 ([[Automation: MoveLayer] Automated transaction prologue/epilogue smr-moonshot#1347](https://github.com/Entropy-Foundation/smr-moonshot/issues/1347)), I believe it is not checked in the prologue that task-id indeed belong to the sender.

Addressed, sender check is also added.

Commit:
Added task sender check in automated_transaction_prologue
